02-client refactor

colin-axner opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Description

This PR contains a large body of work which simplifies the 02-client implementation while also increasing the flexibility for light client implementations. 

Included in these changes include:
- setting client/consensus states within the light client instead of at the 02-client level
- merging header/misbehaviour interface and renaming to `ClientMessage`
- simplification of the solo machine sign bytes
- generic verification functions (VerifyMembership, VerifyNonMembership)
- splitting of the `CheckHeaderAndUpdateState` into 4 functions
- emission of upgraded consensus state

Please see #1846 for more detailed information. This refactor will be included in v6 and a tracking issue will be released soon. 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

seantking commented on CHANGELOG.md:
```diff
@@ -152,6 +174,11 @@ Ref: https://keepachangelog.com/en/1.0.0/
 
 ### API Breaking
 
+* (02-client) [\#598](https://github.com/cosmos/ibc-go/pull/598) The client state and consensus state return value has been removed from `VerifyUpgradeAndUpdateState`. Light client implementations must update the client state and consensus state after verifying a valid client upgrade. 
```
Should this be under v3.0.0?

crodriguezvega replied:
And these are added to the v3.0.0 release...

colin-axner commented on modules/core/03-connection/keeper/handshake_test.go:
```diff
@@ -7,7 +7,7 @@ import (
 	host "github.com/cosmos/ibc-go/v5/modules/core/24-host"
 	"github.com/cosmos/ibc-go/v5/modules/core/exported"
-	ibctmtypes "github.com/cosmos/ibc-go/v5/modules/light-clients/07-tendermint/types"
+	ibctmtypes "github.com/cosmos/ibc-go/v5/modules/light-clients/07-tendermint"
```
I'm curious what we should do with this alias. We currently use `tendermint` in most places, but maybe we should use `ibctm`?

colin-axner commented on docs/roadmap/roadmap.md:
```diff
@@ -14,7 +14,7 @@ This roadmap should be read as a high-level guide, rather than a commitment to s
 
 At a high level we will focus on:
 
-### Features
```
revert

seantking commented on testing/simapp/app.go:
```diff
@@ -38,6 +38,7 @@ import (
 	"github.com/cosmos/cosmos-sdk/x/capability"
 	capabilitykeeper "github.com/cosmos/cosmos-sdk/x/capability/keeper"
 	capabilitytypes "github.com/cosmos/cosmos-sdk/x/capability/types"
+
```
```suggestion
```

charleenfei commented on modules/core/02-client/keeper/events.go:
```diff
@@ -92,15 +115,13 @@ func EmitSubmitMisbehaviourEvent(ctx sdk.Context, clientID string, clientState e
-			sdk.NewAttribute(types.AttributeKeyClientType, clientState.ClientType()),
-			sdk.NewAttribute(types.AttributeKeyConsensusHeight, consensusHeight.String()),
-			sdk.NewAttribute(types.AttributeKeyHeader, headerStr),
+			types.EventTypeUpgradeChain,
```
does the ConsensusHeight still need to be emitted here?

colin-axner replied:
No, not for `EmitSubmitMisbehaviourEventOnUpdate`. That information doesn't really help since there's no action to take when the client becomes frozen. The misbehaviour used to freeze the client can always be queried from the tx 

For `EmitUpgradeChainEvent` the important information is the plan height as that is where the consensus state is stored

charleenfei commented on modules/core/02-client/types/events.go:
```diff
@@ -12,9 +12,11 @@ const (
 	AttributeKeySubjectClientID   = "subject_client_id"
 	AttributeKeyClientType        = "client_type"
 	AttributeKeyConsensusHeight   = "consensus_height"
```
should we add a deprecation notice on this attribute key `AttributeKeyConsensusHeight`?

charleenfei commented on modules/core/exported/client.go:
```diff
@@ -56,10 +52,24 @@ type ClientState interface {
+	// Upon successful update, a list of consensus heights is returned. It assumes the ClientMessage has already been verified.
+	UpdateState(sdk.Context, codec.BinaryCodec, sdk.KVStore, ClientMessage) []Height
+
+	// Update and Misbehaviour functions
```
this line can maybe be removed

charleenfei commented on modules/core/genesis_test.go:
```diff
@@ -26,7 +25,6 @@ const (
 	clientID      = "07-tendermint-0"
 	connectionID2 = "connection-1"
 	clientID2     = "07-tendermin-1"
```
```suggestion
	clientID2     = "07-tendermint-1"
```

charleenfei commented on modules/light-clients/07-tendermint/header_test.go:
```diff
@@ -7,7 +7,7 @@ import (
 	clienttypes "github.com/cosmos/ibc-go/v5/modules/core/02-client/types"
 	"github.com/cosmos/ibc-go/v5/modules/core/exported"
-	"github.com/cosmos/ibc-go/v5/modules/light-clients/07-tendermint/types"
+	tendermint "github.com/cosmos/ibc-go/v5/modules/light-clients/07-tendermint"
```
ibctm?

charleenfei commented on modules/light-clients/07-tendermint/genesis_test.go:
```diff
@@ -1,10 +1,10 @@
 
 	clienttypes "github.com/cosmos/ibc-go/v5/modules/core/02-client/types"
-	"github.com/cosmos/ibc-go/v5/modules/light-clients/07-tendermint/types"
+	tendermint "github.com/cosmos/ibc-go/v5/modules/light-clients/07-tendermint"
```
ibctm?

crodriguezvega commented on CHANGELOG.md:
```diff
@@ -163,6 +190,7 @@ Ref: https://keepachangelog.com/en/1.0.0/
 * (transfer) [\#517](https://github.com/cosmos/ibc-go/pull/517) Separates the ICS 26 callback functions from `AppModule` into a new type `IBCModule` for ICS 20 transfer.
 * (modules/core/02-client) [\#536](https://github.com/cosmos/ibc-go/pull/536) `GetSelfConsensusState` return type changed from bool to error.
 * (channel) [\#644](https://github.com/cosmos/ibc-go/pull/644) Removes `CounterpartyHops` function from the ChannelKeeper.
+* (modules/core/exported) [\#1107](https://github.com/cosmos/ibc-go/pull/1107) Merging the `Header` and `Misbehaviour` interfaces into a single `ClientMessage` type
```
Also in v3.0.0

crodriguezvega commented on CHANGELOG.md:
```diff
@@ -174,6 +202,7 @@ Ref: https://keepachangelog.com/en/1.0.0/
 
 ### Improvements
 
+* (client) [\#888](https://github.com/cosmos/ibc-go/pull/888) Add `GetTimestampAtHeight` to `ClientState`
```
Also in v3.0.0.

crodriguezvega reviewed (commented):
Looks like the changelog is a bit messed up...


Would it be possible to add all the entries for 02-client refactor at the end of each section and separated maybe with a extra line or something? So that it's easier to clean up `Unreleased` when we cut v4.

seantking commented:
> 

I noticed this as well and was wondering what our strategy should be here. 

I think we can open a separate PR to fix the changelog. I can do it :hand:

damiannolan commented on modules/core/migrations/v5/migrations.go:
```diff
@@ -0,0 +1,44 @@
+const Localhost string = "09-localhost"
+
+// MigrateToV5 prunes the 09-Localhost client and associated consensus states from the ibc store
+func MigrateToV5(ctx sdk.Context, clientKeeper clientkeeper.Keeper) {
```
Just flagging as PSA but this will go `v6` as part of upcoming migrations work 👍

damiannolan reviewed (approved):
LGTM! Thanks for handling all the conflicts on this and pushing to get it merged asap! 🚀 
I think everything looks good, but we should also be thorough with the `v6` security audit to make sure nothing has been missed!

colin-axner commented:
I'll pickup all the non-changelog suggested changes :raised_back_of_hand:

colin-axner commented on modules/core/exported/client.go:
```diff
@@ -56,10 +52,24 @@ type ClientState interface {
-	// Update and Misbehaviour functions
+	// VerifyClientMessage must verify a ClientMessage. A ClientMessage could be a Header, Misbehaviour, or batch update.
+	// It must handle each type of ClientMessage appropriately. Calls to CheckForMisbehaviour, UpdateState, and UpdateStateOnMisbehaviour
+	// will assume that the content of the ClientMessage has been verified and can be trusted. An error should be returned
+	// if the ClientMessage fails to verify.
+	VerifyClientMessage(ctx sdk.Context, cdc codec.BinaryCodec, clientStore sdk.KVStore, clientMsg ClientMessage) error
+
+	// Checks for evidence of a misbehaviour in Header or Misbehaviour type. It assumes the ClientMessage
+	// has already been verified.
+	CheckForMisbehaviour(ctx sdk.Context, cdc codec.BinaryCodec, clientStore sdk.KVStore, msg ClientMessage) bool
+
+	// UpdateStateOnMisbehaviour should perform appropriate state changes on a client state given that misbehaviour has been detected and verified
+	UpdateStateOnMisbehaviour(ctx sdk.Context, cdc codec.BinaryCodec, clientStore sdk.KVStore, clientMsg ClientMessage)
 
-	CheckHeaderAndUpdateState(sdk.Context, codec.BinaryCodec, sdk.KVStore, Header) (ClientState, ConsensusState, error)
-	CheckMisbehaviourAndUpdateState(sdk.Context, codec.BinaryCodec, sdk.KVStore, Misbehaviour) (ClientState, error)
+	// UpdateState updates and stores as necessary any associated information for an IBC client, such as the ClientState and corresponding ConsensusState.
+	// Upon successful update, a list of consensus heights is returned. It assumes the ClientMessage has already been verified.
+	UpdateState(sdk.Context, codec.BinaryCodec, sdk.KVStore, ClientMessage) []Height
```
Sounds good to me! the types are pretty self descriptive

colin-axner replied:
All expect `CheckSubstituteAndUpdateState`

colin-axner replied:
On second thought, I think we should just give all the arguments var names so as not to lead to any confusion

damiannolan replied:
That works too!

colin-axner commented on modules/light-clients/07-tendermint/misbehaviour.go:
```diff
@@ -130,3 +125,9 @@ func validCommit(chainID string, blockID tmtypes.BlockID, commit *tmproto.Commit
 	return nil
 }
+
+// TODO: Remove GetHeight()
+// GetHeight implements the curret exported.Header interface, to be updated
+func (misbehaviour Misbehaviour) GetHeight() exported.Height {
+	return nil
+}
```
let's do in a post merge followup

colin-axner replied:
#1881

colin-axner commented on modules/light-clients/07-tendermint/consensus_state_test.go:
```diff
@@ -1,68 +1,58 @@
-		{
-			"success",
-			&types.ConsensusState{
+		{"success",
+			&ibctm.ConsensusState{
```
needs linting

colin-axner commented on modules/core/04-channel/keeper/packet.go:
```diff
@@ -71,7 +71,7 @@ func (k Keeper) SendPacket(
 	clientStore := k.clientKeeper.ClientStore(ctx, connectionEnd.GetClientID())
 	if status := clientState.Status(ctx, clientStore, k.cdc); status != exported.Active {
-		return sdkerrors.Wrapf(clienttypes.ErrClientNotActive, "cannot send packet using client (%s) with status %s", connectionEnd.GetClientID(), status)
+		return sdkerrors.Wrapf(clienttypes.ErrClientStateNotFound, "cannot send packet using client (%s) with status %s", connectionEnd.GetClientID(), status)
```
Seems like this shouldn't have been changed?

colin-axner commented on modules/core/03-connection/keeper/verify.go:
```diff
@@ -28,12 +32,30 @@ func (k Keeper) VerifyClientState(
 
 	if status := targetClient.Status(ctx, clientStore, k.cdc); status != exported.Active {
-		return sdkerrors.Wrapf(clienttypes.ErrClientNotActive, "client (%s) status is %s", clientID, status)
+		return sdkerrors.Wrapf(clienttypes.ErrClientStateNotFound, "client (%s) status is %s", clientID, status)
```
ditto?

colin-axner commented on modules/core/03-connection/keeper/verify.go:
```diff
@@ -28,12 +32,30 @@ func (k Keeper) VerifyClientState(
-	if err := targetClient.VerifyClientState(
-		clientStore, k.cdc, height,
-		connection.GetCounterparty().GetPrefix(), connection.GetCounterparty().GetClientID(), proof, clientState); err != nil {
+	merklePath := commitmenttypes.NewMerklePath(host.FullClientStatePath(connection.GetCounterparty().GetClientID()))
+	merklePath, err := commitmenttypes.ApplyPrefix(connection.GetCounterparty().GetPrefix(), merklePath)
```
there shouldn't be an issues with reusing the same var and `:=` right?

damiannolan replied:
yep, the second `:=` short-hand assignment should be for initialising the `err` return and re-assign the `merklePath`

colin-axner commented on modules/core/03-connection/keeper/verify.go:
```diff
@@ -28,12 +32,30 @@ func (k Keeper) VerifyClientState(
+		return err
+	}
+
+	path, err := k.cdc.Marshal(&merklePath)
+	if err != nil {
+		return err
+	}
```
I think we should remove the marshal/unmarshal and pass the merkle path type directly (that's what the spec specifies as well). We can do in a followup

colin-axner replied:
#1884

colin-axner commented on modules/light-clients/06-solomachine/client_state.go:
```diff
@@ -0,0 +1,264 @@
+// ZeroCustomFields returns solomachine client state with client-specific fields FrozenSequence,
+// and AllowUpdateAfterProposal zeroed out
+func (cs ClientState) ZeroCustomFields() exported.ClientState {
+	return NewClientState(
+		cs.Sequence, cs.ConsensusState, false,
+	)
```
Hmm, technically we shouldn't modify the `IsFrozen` value

colin-axner replied:
#1885

colin-axner commented on modules/light-clients/06-solomachine/update.go:
```diff
@@ -0,0 +1,141 @@
+	cs.Sequence++
+	cs.ConsensusState = consensusState
+
+	clientStore.Set(host.ClientStateKey(), clienttypes.MustMarshalClientState(cdc, &cs))
```
we should have a `setClientState` function

colin-axner replied:
#1887

damiannolan replied:
We have one in [`client_state.go`](https://github.com/cosmos/ibc-go/blob/updated-02-client-refactor/modules/light-clients/06-solomachine/client_state.go#L261). 

My bad on forgetting to use it here!

damiannolan replied:
https://github.com/cosmos/ibc-go/pull/1891

colin-axner commented on modules/light-clients/07-tendermint/update.go:
```diff
@@ -239,28 +110,162 @@ func checkValidity(
+		return true
+	}
+
+	if err := IterateConsensusStateAscending(clientStore, pruneCb); err != nil {
```
we should remove the `error` return from `IterateConsensusStateAscending`

colin-axner commented on modules/core/02-client/types/errors.go:
```diff
@@ -33,5 +33,5 @@ var (
 	ErrInvalidSubstitute                      = sdkerrors.Register(SubModuleName, 27, "invalid client state substitute")
 	ErrInvalidUpgradeProposal                 = sdkerrors.Register(SubModuleName, 28, "invalid upgrade proposal")
-	ErrClientNotActive                        = sdkerrors.Register(SubModuleName, 29, "client is not active")
+	ErrClientStateNotFound                    = sdkerrors.Register(SubModuleName, 29, "client state not found")
 )
```
weird

colin-axner commented on modules/light-clients/06-solomachine/update.go:
```diff
@@ -0,0 +1,141 @@
+		Sequence:    header.Sequence,
+		Timestamp:   header.Timestamp,
+		Diversifier: cs.ConsensusState.Diversifier,
+		Path:        []byte{},
```
cc @AdityaSripal

damiannolan replied:
I wondered this myself when changing around the code. Maybe we could at least define a constant of some sort if we don't use a specific path

colin-axner replied:
opened #1915 and #1916

colin-axner commented on modules/light-clients/06-solomachine/client_state.go:
```diff
@@ -0,0 +1,264 @@
+		return nil, nil, 0, 0, err
+	}
+
+	if cs.ConsensusState == nil {
+		return nil, nil, 0, 0, sdkerrors.Wrap(clienttypes.ErrInvalidConsensus, "consensus state cannot be empty")
+	}
```
#1888

colin-axner commented:
I've done a self review, addressed all issues that required immediate action and opened up issues for the others. My remaining questions are commented above. Specifically interested in construction of sign bytes for header/verify non membership

damiannolan commented on modules/light-clients/06-solomachine/client_state.go:
```diff
@@ -0,0 +1,264 @@
+		Timestamp:   timestamp,
+		Diversifier: cs.ConsensusState.Diversifier,
+		Path:        []byte(merklePath.String()),
+		Data:        nil,
```
I think it would be good to test this but I'd imagine there is a difference

colin-axner replied:
Opened #1914

AdityaSripal commented on CHANGELOG.md:
```diff
@@ -94,6 +120,10 @@ Ref: https://keepachangelog.com/en/1.0.0/
 * (apps/29-fee) [\#1278](https://github.com/cosmos/ibc-go/pull/1278) The URI path for the query to get all incentivized packets for a specific channel did not follow the same format as the rest of queries.
+-
+* (light-clients/solomachine) [#1839](https://github.com/cosmos/ibc-go/issues/1839) Fixed usage of the new diversifier in validation of changing diversifiers for the solo machine. The current diversifier must sign over the new diversifier.
+* (07-tendermint) [\#1674](https://github.com/cosmos/ibc-go/pull/1674) Submitted ClientState is zeroed out before checking the proof in order to prevent the proposal from containing information governance is not actually voting on.
```
```suggestion
* (light-clients/07-tendermint) [\#1674](https://github.com/cosmos/ibc-go/pull/1674) Submitted ClientState is zeroed out before checking the proof in order to prevent the proposal from containing information governance is not actually voting on.
```

?

AdityaSripal commented on docs/architecture/adr-027-ibc-wasm.md:
````diff
@@ -100,7 +100,7 @@ packaged inside a payload which is json serialized and passed to `callContract`
 
 ```go
-func (c *ClientState) CheckProposedHeaderAndUpdateState(context sdk.Context, marshaler codec.BinaryMarshaler, store sdk.KVStore, header exported.Header) (exported.ClientState, exported.ConsensusState, error) {
+func (c *ClientState) CheckProposedHeaderAndUpdateState(context sdk.Context, marshaler codec.BinaryMarshaler, store sdk.KVStore, header exported.ClientMessage) (exported.ClientState, exported.ConsensusState, error) {
````
Shouldn't this be split up into two functions or we will allow the ADR authors to update?

AdityaSripal commented on docs/migrations/v5-to-v6.md:
```diff
@@ -0,0 +1,101 @@
+
+- `UpdateState` updates and stores as necessary any associated information for an IBC client, such as the `ClientState` and corresponding `ConsensusState`. An error is returned if `ClientMessage` is of type `Misbehaviour`. Upon successful update, a list containing the updated consensus state height is returned.
+
+The `CheckMisbehaviourAndUpdateState` function has been removed from `ClientState` interface. This functionality is now encapsulated by the usage of `VerifyClientMessage`, `CheckForMisbehaviour`, `UpdateStateOnMisbehaviour`, `UpdateState`.
```
```suggestion
The `CheckMisbehaviourAndUpdateState` function has been removed from `ClientState` interface. This functionality is now encapsulated by the usage of `VerifyClientMessage`, `CheckForMisbehaviour`, `UpdateStateOnMisbehaviour`.
```

UpdateState doesn't get used here correct?

AdityaSripal reviewed (approved):
Amazing work team 🚀
